requests-for-comments/2021-10-05-group-analytics.md

```diff
@@ -0,0 +1,196 @@
+
+## Overview
+
+This RFC proposes a solution which would allow PostHog users to do analysis not only by users but also by groups (e.g. organizations, projects).
+
+Let's make it possible 
+
+## Background
+B2B products don’t deal directly with users, but with other companies and hence have more complicated needs around what kinds of metrics are most important for them.
+
+
+Lack of this feature has been a major driver for users wanting data exports in different analytics products, forcing users to roll their own analytics on top of data warehouses to get at their key metrics.
+
+### Example usecases
+
+
+- Show me how many active “organizations” there are
+- Show only pageviews from organizations who are paying us (filtering by group properties)
+- Have an overview of groups and their relations with users
+- Feature flags: roll out feature group-by-group
+
+For PostHog the product, the 3 groups we would use are:
+
+- Instances (e.g. cloud, posthog.someproduct.org, …)
+- Organizations
+- Projects
+
+## Semantics & Constraints
+
+### 1. Users can belong to multiple groups
+
+Some examples:
+
+1. Contractor type work (e.g. for farm management agronomists work for multiple farms at once)
+2. Billing happens on a different level (e.g. our projects vs organization)
+3. Users clients are global companies with a lot of local “branches” where execs need access to everything
+4. Internal users have access to debug issues on multiple accounts
+### 2. User is active in only “group” at any point in time
+
+Whenever an event is made, we can (and should) make the assumption that it belongs to a single group (e.g. organization). This is needed to avoid “double-counting” in stats.
+
+Example: If Yakko has been added to 20 different organizations and is active, we don’t want to count all of them as active.
+
+### 3. Limited amount of groups
+
+From speaking with clients, their internal amount of different “groups” is limited. 
+
+This constraint allows us to implement things slightly more efficiently and have a nicer UX compared to unbounded amount of groups.
+
+### 4. Feature flags in groups should be consistent
+
+For the users who are using feature flags, they want to ensure a consistent experience within a group during rollout. 
+
+This is to avoid situations where some users see a new feature and some do not.
+
+This also factors into their later analysis - for example, they’d like to treat two different “organizations” both as a single datapoint even if one has 1500 users and the other has 1.
+
+## Proposal
+
+### Groups are event-level
+
+Users can define up to N (5?) group types. For us these would be `instance` , `project` and `organization`.
+
+Each event we ingest can have one group per group type associated. 
+
+We will have N new columns in the `events` table, containing the group ids - first one for the first group type of the project, second for the second one etc. 
```

Comment:
I suspect you've thought long & hard about why we're doing this on the events table and not on persons. Would be nice to understand why you've come to this solution over doing it person based.

Reply:
I hoped it was already clear based on the constraints but added a section at the end under `Other details`.